rust/otel-arrow-rust/src/otlp/attributes/decoder.rs

```diff
@@ -62,3 +76,532 @@ where
         }
     }
 }
+
+/// Decodes the parent IDs from their transport optimized encoding to the actual ID values.
+///
+/// In the transport optimized encoding, the record batch is sorted by value Type, then
+/// attribute key, then attribute value. If subsequent rows have the same key & value, the
+/// parent IDs are delta encoded.
+///
+/// There are additional exceptions that end the sequence of delta encoding:
+/// - null values (we don't consider null values equal, even if current row & previous
+///   row value are both null)
+/// - value types Empty, Slice & Map are never considered equal
+///
+/// For example:
+///
+/// | attr key |  attr val  | parent_id |
+/// |----------|------------| --------- |
+/// |   "a1"   |  str("a")  |  0        | <-- parent id = 0
+/// |   "a1"   |  str("a")  |  0        | <-- key & val == previous row -> delta encoded -> parent id = prev id + 0 = 0
+/// |   "a1"   |  str("a")  |  1        | <-- key & val == previous row -> delta encoded -> parent id = prev id + 1 = 1
+/// |   "a1"   |  str("a")  |  2        | <-- key & val == previous row -> delta encoded -> parent id = prev id + 2 = 3
+/// |   "a1"   |  str("b")  |  0        | <-- val != prev val -> no delta encoding -> parent id = 0
+/// |   "a2"   |  str("b")  |  0        | <-- key != prev val -> no delta encoding -> parent id = 0
+/// ...
+/// |   "a3"   | slice([0]) |  0        | <-- parent id = 0
+/// |   "a3"   | slice([0]) |  0        | <-- val type = slice -> no delta encoding -> parent id = 0
+/// ...                                       (^ it would be same if value type = Map or Empty)
+/// ...
+/// |   "a4"   | str(null) |  0        | <-- parent id = 0
+/// |   "a4"   | str(null) |  0        | <-- value is null -> no delta encoding even though key + value are same
+///
+/// This returns a new RecordBatch with the parent_id column replaced with the materialized id.
+///
+#[allow(unused)] // TODO -- remove allow(unused) when we use this to optimize decoding OTAP
+pub fn materialize_parent_id<T>(record_batch: &RecordBatch) -> Result<RecordBatch>
+where
+    T: ParentId,
+    <T as ParentId>::ArrayType: ArrowPrimitiveType,
+    <<T as ParentId>::ArrayType as ArrowPrimitiveType>::Native: AddAssign,
+{
+    // if the batch is empty, just skip all this logic and return a batch
+    if record_batch.num_rows() == 0 {
+        return Ok(record_batch.clone());
```

Comment:
It looks like the arrow crate offers a method to create an empty `RecordBatch`. It might be good to check if using that would be more performant. https://arrow.apache.org/rust/arrow/array/struct.RecordBatch.html#method.new_empty
```suggestion
        return Ok(RecordBatch::new_empty(record_batch.schema()));
```

Reply:
I'm not sure if `empty_batch` is more optimized than `clone`. 

It looks like `RecordBatch::new_empty` creates a new vec of empty arrays for each field in the schema, and in doing so creates new `Arc` for each column). By contrast, `RecordBatch::clone` will create will clone the Arc for the schema, then create a new vec and fill it with cloned `Arc` for each exiting empty column. 

I thought cloning an Arc::clone was generally faster than Arc::new, so I'd expect RecordBatch::clone to be faster than new_empty.

I'm planning on writing a benchmark for this tomorrow as suggested [here](https://github.com/open-telemetry/otel-arrow/pull/455#pullrequestreview-2851663866), so I could test this out to confirm either way

Reply:
@utpilla using RecordBatch::new_empty here is quite a bit slower:

Current code path calling `record_batch.clone()`
```
$ RUST_BACKTRACE=full  cargo bench --bench materialize_parent_id
# ...
materialize_parent_ids/materialize_parent_ids/0
                        time:   [21.644 ns 21.675 ns 21.707 ns]
```
```
$ RUST_BACKTRACE=full  cargo bench --bench materialize_parent_id
# ...
materialize_parent_ids/materialize_parent_ids/0
                        time:   [1.4948 µs 1.4995 µs 1.5049 µs]
                        change: [+6776.3% +6803.6% +6828.3%] (p = 0.00 < 0.05)
                        Performance has regressed.
  ```

Reply:
It seems off if the only major difference between the two is about `Arc:new` vs `Arc::clone`. From 20ns to 1.5 us. That's almost 75 times more!

Reply:
I think the difference can be explained by the fact that a clone only duplicates the Arcs, whereas the empty call not only creates the Arcs but also all the data structures pointed to by those Arcs.